### ISSUE TITLE
SharpNL is not thread safe

### ISSUE THREAD

**Issue by JobaDiniz** [knuppe/SharpNL#42]
Cache class https://github.com/knuppe/SharpNL/blob/development/src/SharpNL/Utility/Cache.cs is not thread safe... 
We could use `ConcurrentDictionary` to handle this... is this a change that we can make?

**Comment by knuppe** [comment]
@JobaDiniz you're right, this class is not thread-safe. But the library follows the same pattern as opennlp, templates are thread safe, and tools are not.

for example, you can share `SentenceModel` (the most expensive part to load) between multiple threads, but `SentenceDetectorME` must be used by a single thread.

Since this is a port I decided not to change the architecture too much.

**Comment by JobaDiniz** [comment]
So, in a multi-thread env, the idea is to always instantiate a new `SentenceDetectorME` using the same `SentenceModel`? Instanciating `SentenceDetectorME` is not a heavy operation?

**Comment by TechnikEmpire** [comment]
@JobaDinzi just implement a lock. If you can demonstrate that this is a bottleneck to your wider application I'd be very interested in seeing that.

**Comment by TechnikEmpire** [comment]
If it really is a bottleneck (locking), then implement a thread pool with per-thread instances.
